sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala

```diff
@@ -1549,6 +1551,38 @@ class DatasetSuite extends QueryTest with SharedSQLContext {
       Seq(Row("Amsterdam")))
   }
 
+  test("SPARK-25951: avoid redundant shuffle on rename") {
```

Comment:
I'm a little worried about introducing such a big change for such a corner case.

Looking at this test case, it seems updating the `outputPartitioning` according to the project list is good enough?

cc @maryannxue

Reply:
> Looking at this test case, it seems updating the outputPartitioning according to the project list is good enough?

Basically, that is what is done. The diff is quite big, that's true, but mostly are tests. Is there something specific you are worried about?

Reply:
I'm worried about `Expression.sameResult`. Is it still useful after updating the outputPartitioning according to the project list?

Reply:
it is still useful because it may happen a used writes something like:
```
select ... group by a as b, ...
```
(please notice that despite it makes no sense written in SQL, maybe with other APIs is easier to happen).

And especially it is useful when updating the `outputPartitioning` in order to check whether an expression produces the same result of another.